Sort `list_observation_ids()`, support importing first or last observation ID from a volume file

nilsvu opened the pull request:
## Proposed changes

Now you can specify `ObservationValue: Last` to load initial data, so the observation value doesn't have to be known.

Also:

- Sort `VolumeData::list_observation_ids()`.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Sort observation IDs

nilsdeppe commented on src/IO/H5/VolumeData.cpp:
```diff
@@ -273,8 +273,15 @@ std::vector<size_t> VolumeData::list_observation_ids() const {
+  std::vector<size_t> obs_ids{
+      boost::make_transform_iterator(names.begin(), helper),
+      boost::make_transform_iterator(names.end(), helper)};
+  std::sort(obs_ids.begin(), obs_ids.end(),
```
`alg::sort(obs_ids, ...);`

nilsvu replied:
squashed this in

nilsdeppe commented on src/IO/Importers/Actions/ReadVolumeData.hpp:
```diff
@@ -5,6 +5,7 @@
 
 #include <cstddef>
```
Can you split this commit? I'm not sure how many different things it's doing at once and am having a hard time following how to parse things... This should be at least 1 commit adding the selector enum, 1 commit using it, and possibly more commits changing target link libraries and `std::get`s.

nilsvu replied:
split the commits 👍

nilsvu committed:
Add importers::ObservationSelector enum

This is used to select the first or last observation in a volume file,
to allow specifying it in an input file without knowledge of the
specific observation values.

nilsvu committed:
Support importing first or last observation ID from a volume file

Now you can specify `ObservationValue: Last` to load initial data, so
the observation value doesn't have to be known.

nilsvu committed:
Fix linking of Test_DataImporter

nilsdeppe commented on src/IO/Importers/ObservationSelector.cpp:
```diff
@@ -0,0 +1,41 @@
+#include "Utilities/ErrorHandling/Error.hpp"
+
+namespace importers {
+std::ostream& operator<<(std::ostream& os, const ObservationSelector& value) {
```
take value by value

nilsvu replied:
fixed

nilsdeppe reviewed (commented):
LGTM, one small note. Please squash (sorry for not catching  before)

nilsvu commented:
Rebase on #3822 wasn't entirely trivial, so I pushed a fixup.

nilsdeppe reviewed (commented):
LGTM, go ahead and squash
